rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2150 @@
+        /// the sum of `wizz` and `bang` must be greater than or equal to zero.
+        pub fn new(wizz: i8, bang: u8) -> Self {
+            assert!((wizz as i16) + (bang as i16) >= 0);
+            Constrained(private::ConstrainedInner { wizz, bang })
+        }
+
+        pub fn something_dangerous(&self) {
+            unsafe {
+                // do something that's only safe if `self.0.wizz + self.0.bang >= 0`
+            }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        /// the sum of `wizz` and `bang` MUST be greater than or equal to zero.
+        #[repr(C)]
+        pub struct ConstrainedInner {
+            pub wizz: i8,
+            pub bang: u8,
+        }
+    }
+
+}
+```
+In the above example, the definitions of both `Constrained` and its field `ConstrainedInner` are marked `pub`. All fields of both types are marked `pub`. However, `Constrained` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `ConstrainedInner` is not visibile outside of `crate_a`.
+
+#### Constructability and Transmutation
+Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author.
+
+In the previous examples, it would be *unsafe* to transmute `0xFF00u16` into `Constrained` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `something_dangerous`) could violate memory safety. (However, it's completely safe to transmute `Constrained` into `0xFF00u16`.)
+
+
+For transmutations where the destination type involves mutate-able references, the constructability of the source type is also relevant. Consider:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let mut x = Constrained::new(0, 0);
+{
+    let y : &mut u16 = (&mut x).transmute_into(); // Compile Error!
+    *y = 0xFF00u16;
+}
+
+let z : Constrained = x;
+```
+If this example did not produce a compile error, the value of `z` would not be a safe instance of its type, `Constrained`.
+
+### 📖 When is a transmutation stable?
+[stability]: #When-is-a-transmutation-stable
+
+Since the soundness and safety of a transmutation is affected by the layouts of the source and destination types, changes to those types' layouts may cause code which previously compiled to produce errors. In other words, transmutation causes a type's layout to become part of that type's API for the purposes of SemVer stability.
+
+The question is, then: *how can the author of a type reason about transmutations they did not write, from-or-to types they did not write?* We address this problem by introducing two traits which both allow an author to opt-in to stability guarantees for their types, and allow third-parties to reason at compile-time about what guarantees are provided for such types.
+
+#### `PromiseTransmutableFrom` and `PromiseTransmutableInto`
```

Comment:
I'm fond of these names because `Promise` reflects that the user implements these traits to *make* a promise, and `Transmutable{From,Into}` corresponds exactly to what they're promising. I'm quite open to bikeshedding, though!